### ISSUE TITLE
Going in reverse 

### ISSUE THREAD

**Issue by fb1812** [prl-mushr/mushr#116]
Hi, I want the car to switch behavior from going forward to going in reverse based on the path that I give. Right now, the car goes all the time forward and I am not sure how to switch to the reverse behavior automatically based on the path the controller receives. 

Would you be able to help me with this? 

Many thanks for considering my request.

**Comment by sidtalia** [comment]
Hi!

You may using an older version of the mushr_rhc. In the old version, only forward velocities were being sampled. In the noetic branch, it should sample forward, 0, and reverse. [This ](https://github.com/prl-mushr/mushr_rhc/blob/noetic/src/mpc.py#L58)is how it is currently implemented.

If you wish to not change branches, you can take the logic used in the noetic branch and sample forward, 0 and reverse actions. 

Hope this helps.

Thanks,
Regards,
Sid.

**Comment by fb1812** [comment]
Not sure I am following. I have noetic branch and mushr_base and then I have my own controller implemented that is just a simple controller to follow the path that I am giving. I am not s sure if it supposed to just go in reverse automatically if I give negative velocity and backward waypoints (and right now this is not happening) and I am not sure if the problem is mine or if there is something I need to set up in the vesc  to make the reverse maneuver happening

**Comment by sidtalia** [comment]
I see. If you're doing this on a real car, could you place it on a block such that the tires don't touch the ground, and hardcode the commanded velocity in your controller to always be negative? If the wheels move backwards, it may be a bug in your code.

Thanks,
Regards,
Sid

**Comment by DKrishna007** [comment]
Yeah, you're absolutely right — by default, the car is set up to always drive forward. To get it to switch between forward and reverse automatically based on the path, you’ll need to tweak a couple of things.

Here's what you can look into:
Controller settings
Some controllers (like mpc_controller) may have a parameter that restricts motion to forward only. Check if there’s a flag related to reverse or velocity limits, and make sure reverse (i.e., negative velocity) is allowed.

Path planning logic
If you’re generating your own paths, make sure the velocity values actually go negative when the path makes more sense in reverse. Sometimes it's just a matter of flipping the sign depending on the desired heading.

Controller logic
In the controller code (maybe something like control_law.py), make sure it handles negative velocity properly. This might also mean flipping the steering angle or adjusting how turning is handled in reverse.

Bonus
If you're manually sending waypoints, you could include a flag or tag in your path to indicate that a segment should be done in reverse — and then adjust your control accordingly.

Let me know if you want help finding the exact files or functions where these changes happen — happy to point you to them!
